packages/suite/src/utils/suite/messageSystem.ts

```diff
@@ -0,0 +1,221 @@
+
+    if (typeof version === 'string') {
+        return version;
+    }
+
+    return version.join(' || ');
+};
+
+export const validateDurationCompatibility = (durationCondition: Duration): boolean => {
+    const currentDate = Date.now();
+
+    const from = Date.parse(durationCondition.from);
+    const to = Date.parse(durationCondition.to);
+
+    return from <= currentDate && currentDate <= to;
+};
+
+export const validateVersionCompatibility = (
+    condition: Browser | OperatingSystem | Environment | Transport,
+    type: string | SuiteEnvironmentType,
+    version: string,
+): boolean => {
+    const conditionVersion = normalizeVersion(condition[type]);
+
+    if (conditionVersion === null) {
+        return false;
+    }
+
+    return semver.satisfies(version, conditionVersion);
+};
+
+export const validateSettingsCompatibility = (
+    settingsCondition: Settings[],
+    currentSettings: CurrentSettings,
+): boolean => {
+    const settings: {
+        [key: string]: any;
+    } = currentSettings.enabledNetworks.reduce((o, key) => Object.assign(o, { [key]: true }), {
+        tor: currentSettings.tor,
+    });
+
+    return settingsCondition.some(settingCondition =>
+        Object.entries(settingCondition).every(
+            ([key, value]: [string, boolean | unknown]) =>
+                settings[key] === value || (!value && settings[key] === undefined),
+        ),
+    );
+};
+
+export const validateTransportCompatibility = (
+    transportCondition: Transport,
+    transport?: Partial<TransportInfo>,
+): boolean => {
+    if (!transport || !transport.type || !transport.version) {
+        return false;
+    }
+
+    const { version } = transport;
+    const type = transport.type.toLowerCase();
+
+    return validateVersionCompatibility(transportCondition, type, version);
+};
+
+export const validateDeviceCompatibility = (
+    deviceConditions: Device[],
+    device?: TrezorDevice,
+): boolean => {
+    // if conditions are empty, then device should be empty
+    if (!deviceConditions.length) {
+        return !device;
+    }
+    if (!device || !device.features) {
+        return false;
+    }
+
+    const {
+        model,
+        vendor,
+        major_version: majorVersion,
+        minor_version: minorVersion,
+        patch_version: patchVersion,
+    } = device.features;
+
+    const deviceVersion = `${majorVersion}.${minorVersion}.${patchVersion}`;
+
+    return deviceConditions.some(
+        deviceCondition =>
+            deviceCondition.model.toLowerCase() === model.toLowerCase() &&
+            deviceCondition.vendor.toLowerCase() === vendor.toLowerCase() &&
+            semver.satisfies(deviceVersion, normalizeVersion(deviceCondition.firmware)!),
+    );
+};
+
+export const getValidMessages = (config: MessageSystem | null, options: Options): Message[] => {
+    if (!config) {
+        return [];
+    }
+
+    const { device, transport, settings } = options;
+
+    const ua = Bowser.getParser(getUserAgent());
```

Comment:
I will replace bowser with ua-parser.js which is used in browser-detection to have it united. 

The solution we talked about is not applicable for the desktop app as browser detection is not available there.

Having an in-house detection module requires a lot of regexes for parsing user agent to correctly detect os and browser versions. This would add a lot of painful code, difficult to maintain, to our project.

Reply:
+ I will create helper functions in env.ts file as we discussed